### ISSUE TITLE
Brush is offset from cursor

### ISSUE THREAD

**Issue by andersmmg** [Zylann/godot_heightmap_plugin#142]
**Describe the bug**
The brush circle if offset from where my cursor is, and it's further off the further I am from the terrain. Makes editing it really hard, especially since the offset isn't always the same so drawing something like a circle ends up as a sort of tilted oval. I've had the problem on multiple projects, but haven't been able to reproduce it when I try.

**To Reproduce**
It only happens sometimes, but I still have yet to find a fix. A bit after creating a terrain, one day it'll just be way off.
If you see any errors in the console, that may also help.

**Expected behavior**
The brush should stay aligned with the cursor.

**Screenshots**
![terrainnope](https://user-images.githubusercontent.com/16022203/79029662-8e1cac80-7b52-11ea-865b-ee46f7b45de4.png)

**Environment**
- OS: Linux
- Graphics card: AMD CARRIZO
- Godot version: 3.2.1
- Plugin version: latest from asset library (afaik)
- Renderer used: GLES3

**Comment by Zylann** [comment]
Did you transform the terrain? (scale, translation, rotation?)

**Comment by andersmmg** [comment]
It has happened when I did that, but no I didn't this time or some other
times

**Comment by Wallace99** [comment]
You don't have the editor perspective set to half resolution do you?

**Comment by andersmmg** [comment]
I don't, but I have done that sometimes. Could it affect it if it's not
currently on half resolution now but it has been recently? Though it's also
happened on projects I never used half resolution on.

**Comment by Wallace99** [comment]
I'm not too sure, I know I've had the same thing happen when editor set to half resolution using this plugin: https://github.com/Zylann/godot_scatter_plugin which I assume uses a similar way of placing stuff at the cursor

Strange that it only happens sometimes though

**Comment by Zylann** [comment]
@Wallace99 I fixed the half-resolution thing with the terrain pluging but not with scatter

**Comment by andersmmg** [comment]
So far scatter hasn't done that to me, though it does sometimes place like
3 of the same scene in the same place with one click. It's annoying, but
otherwise scatter rocks

**Comment by tinmanjuggernaut** [comment]
The cursor is offet on Windows when the brush size is 1 or 2. When the player is a size of 1 or 2, that's the size I want to paint at. But at 1 or 2, the cursor bounces all over the place, except where I want it (probably popping between vertices), but it's not aligned with the cursor. It's very annoying. My 1k Terrain is not transformed.

I have not had it as far off as Op, but I have had it non-responsive, especially when in ortho view, but only sometimes.

**Comment by Zylann** [comment]
@tinmanjuggernaut you realize size 1 or 2 is barely a pixel in terms of heightmap resolution, right^^ It bounces because of the pixel grid...
I think the offset shown in the first screenshot is something else entirely.

**Comment by tinmanjuggernaut** [comment]
I didn't realize that. But in game, my 1x2x1 character might walk on a 1-3 width path. That's not a tiny scale. The default cube is 1x1x1. A path or a groove that's 5-10, which is easier to paint, is unnaturally big and the cursor doesn't paint at size 1 very well. Maybe I could scale everything else up.

**Comment by Zylann** [comment]
@tinmanjuggernaut eventually the brush could be made to use interpolation so it can paint "in between" pixels, although it will significantly lower its speed, particularly with the GDScript backend

**Comment by RonanZe** [comment]
Aren't you using a C++ lib now?

**Comment by Zylann** [comment]
@RonanZe I do, but it only runs on Windows 64 bits for Godot 3.2.1 users specifically. And that would also slow it down, although surely much less.

**Comment by RonanZe** [comment]
Can't you use some kind of shader layer paint tech? 
Paint directly to the graphic card they saved on the drive when possible?

The new shader capacity in Godot 4 seems to give some possibilites in that direction:
[https://godotengine.org/article/godot-40-gets-global-and-instance-shader-uniforms](url)

Is there a fast function in Godot to save the texture on the drive?

**Comment by Zylann** [comment]
@RonanZe yes I would like to move to GPU later but that's going to be Godot 4. Godot 3 can somewhat do that too but it calls for way too many changes and issues. It's still not related to the topic though^^"

**Comment by Calinou** [comment]
@Zylann Maybe brush interpolation could be enabled only at very low brush sizes. This would reduce its performance impact on day-to-day use.

**Comment by Gyeff** [comment]
I have also noticed this issue on my low resolution game, I assumed it was because of the low resolution. Mainly occurs with small brush as mentioned already. 

This is not a major problem in my opinion (or even a problem at all really). This is because it's not difficult to work even with the displaced circle indicator. Just pretend like the circle is under the mouse because that's where the painting actually takes place.

I personally wouldn't use 1-2 all the time. Only use it for minor adjustments if necessary. This is because with a brush that small there are some other problems too. For example, with flatten tool on 100% opacity, it doesn't pull the verts all the way up to the flatten level, it pulls verts as if the brush has some opacity.

A thing that you can do with low scale games is try to play with the x scale and z scale of the heightmap. You need to do this anyway because otherwise low scale sprites and objects appear to clip through terrain in certain areas with big height changes. This is because of vert arrangement.

**Comment by Zylann** [comment]
Note that subpixel precision is likely not what @andersmmg has experienced in the first post, because it looks like it happened at a "regular" size (until further info is known?).
If subpixel precision is desired it has to be another issue.
